### ISSUE TITLE
How to add SaveChanges interceptors when using AddSqlServerDbContext

### ISSUE THREAD

**Issue by dedalusmax** [dotnet/aspire#6902]
### Discussed in https://github.com/dotnet/aspire/discussions/4202

<div type='discussions-op-text'>

<sup>Originally posted by **jstynder** May 16, 2024</sup>
When adding interceptors this would be the way I do It now:```

  services.AddDbContextPool<T>((serviceProvider, optionsBuilder) =>
      optionsBuilder
          .UseSqlServer(connectionString)
          .AddInterceptors(serviceProvider.GetRequiredService<SecondLevelCacheInterceptor>()));
```
How can I access the serviceProvider when using AddSqlServerDbContext

          
          </div>

**Comment by jack775544** [comment]
You should be able to use the [EnrichSqlServerDbContext](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.aspiresqlserverefcoresqlclientextensions.enrichsqlserverdbcontext?view=dotnet-aspire-9.0) method to achieve what you want. The docs for this approach are here: https://learn.microsoft.com/en-us/dotnet/aspire/database/sql-server-entity-framework-integration?tabs=dotnet-cli%2Cssms#add-sql-server-database-context-with-enrichment

This way you can add the context using the AddDbContextPool method like you showed above and the enrich method will add all the extra things like telemetry and health checks.

**Comment by DamianEdwards** [comment]
Yep the approach @jack775544 suggests is what we intend here. You need to use the regular `AddDbContextPool` method and then call `EnrichSqlServerDbContext` to add the Aspire goodness.

**Comment by dedalusmax** [comment]
Yes, but this means we cannot use a containerized SQL database, but rather a real one - with the extra things that Aspire provides. This is what bothers me (or probably other guys) - I want to use AddResource like a containerized resource, and on top of that, use everything that is necessary to configure a DB with EF Core (like adding interceptors alike).  

p.s. I find a general problem that all the methods for configuring the integrations are made as extensions to the IHostBuilder - not IServiceCollection - which doesn't fit in all scenarios (like I'm building now with a console app).

**Comment by davidfowl** [comment]
> Yes, but this means we cannot use a containerized SQL database, but rather a real one 

What does this mean? How does using the enrich method change this?

**Comment by dedalusmax** [comment]
Sorry if I'm not clear in describing the problem. For clarity, just refer to the original question/problem that is asked.
This is what is currently possible/intended with Aspire's configuration:

    builder.**AddSqlServerDbContext**<ApplicationDbContext>(connectionName: "sql-db", configureDbContextOptions: **options** =>
    {
        options.AddInterceptors(); // no way to correctly configure it!!!
    });

this is what is needed:

    builder.AddSqlServerDbContext<ApplicationDbContext>(connectionName: "sql-db", **(serviceProvider, options)** =>
    {
        var interceptor = **serviceProvider**.GetService<AuditableEntitiesInterceptor>()!;
        options.UseSqlServer(connectionString).AddInterceptors(interceptor);
    });


P.S. Enrich method won't help, as it takes **MicrosoftEntityFrameworkCoreSqlServerSettings** type as a parameter:
builder.EnrichSqlServerDbContext<ApplicationDbContext>(settings => { });

**Comment by davidfowl** [comment]
You are supposed to use the enrich method with the normal AddDbContext method.

```C#
builder.AddDbContext<ApplicationDbContext>(options =>
{
    options.UseSqlServer(builder.Configuration.GetConnectionString("sql-db"));
    options.AddInterceptors(..);
});

builder.EnrichSqlServerDbContext<ApplicationDbContext>();
```

Maybe we need to improve our docs here @IEvangelist

**Comment by dedalusmax** [comment]
that worked! thank you for the explanation.
yes, i would say it's rather confusing and not clarified completely.

kind regards!
